core/arch/arm/kernel/stmm_sp.c

```diff
@@ -102,6 +104,39 @@ static void init_stmm_regs(struct stmm_ctx *spc, unsigned long a0,
 	spc->regs.sp = sp;
 	spc->regs.pc = pc;
 }
+#endif
+
+#ifdef ARM32
+static uint32_t __maybe_unused get_spsr(void)
+{
+	uint32_t s;
+
+	s = read_cpsr();
+	s &= ~(CPSR_MODE_MASK | CPSR_T | ARM32_CPSR_IT_MASK);
+	s |= CPSR_MODE_USR;
+	/* StMM 32bit is built in Tumb2 mode */
+	s |= CPSR_T;
+
+	return s;
+}
+
+static void init_stmm_regs(struct stmm_ctx *spc, unsigned long a0,
+			   unsigned long a1, unsigned long sp, unsigned long pc)
+{
+	uint32_t *entry = (void *)pc;
+
+	spc->regs.r0 = a0;
+	spc->regs.r1 = a1;
+	spc->regs.usr_sp = sp;
+	spc->regs.cpsr = get_spsr();
+
+	/*
+	 * Hack: force Thumb instruction mode and extract entry address
+	 * from entry instruction expected as a relative branch.
+	 */
+	spc->regs.pc = pc + (*entry & GENMASK_32(23, 0)) * 4 + 8;
```

Comment:
I'd prefer if this hack wasn't needed.
But when reading from user mode memory we should check that it's safe to read.

Reply:
I'm discussing this the edk2 people. This is indeed ugly. I think we agree that image should be entered in Arm instruction set.  Entry sequence should do the work if needing to switch to Thumb at some point.
I'll remove that and hopefully replace with `spc->regs.pc = pc;`